website/blog/2022-10-31-simple-1.mdx

```diff
@@ -0,0 +1,102 @@
+
+In the past year we have been working on several improvements on the simple function interface on both the `expressivity` and `performance` axes that we will discuss in this series of notes. In this blogpost, we will talk about
+some of the general optimizations that we have in the adapter.
+
+## General optimizations
+
+#### Vector reuse
+If the output type matches one of the input types, and the input vector is to die after the function invocation, then it is possible to reuse it for the results instead of allocating a new vector.
+
+#### Bulk null setting
+Nulls are represented in a bit vector, hence, writing each bit can be expensive specially for primitive operations (like plus and minus). One optimization is
+to optimize for the `not null` case, and bulk setting the nulls to not null (64 bits in one instruction). After that during the computation, only if the results are null, the null bit is set to null.
+
+#### Null setting avoidance
+The adapter can statically infer if a function never generates null; In the simple function interface if the `call` function return's type is `void`, it means the output is never null, and if it's `bool`, then the function returns true for not null and false for null).
+When the function does not generate nulls, then null setting is **completely avoided** during the computation (only the previous bull setting is needed). The consequence of that is that the hot loop applying the function becomes simdizable triggering a huge boost in performance for primitive operations.
+
+
+#### Encoding based fast path
+Vectors in Velox can have different encodings (flat, constant..etc). The generic way of reading a vector of arbitrary encoding is to use a decoded vector to guarantee correct data access. That translates into an overhead each time an input value is accessed (we need to check the encoding of the vector to know how to read the value for every row).
+
+When the function is a primitive operation like plus or minus, such overhead is expensive! To avoid that, encoding based fast paths can be added, the code snippet below illustrates the idea.
+<figure>
+<img src="/img/simple1_11.png" width="60%"/>
+</figure>
+
+In the code above, the overhead of checking the encoding is switched outside the loop that applies the functions (the plus operation here). And the inner loops are simple operations that are potentially simdizable and free of encoding checks.
+One issue with this optimization is that the core loop is replicated many times. In general, the numbers of times it will be replicated
+is `n^m` where `n` is the number of args, and `m` is the number of encodings.
+
+To avoid code size blowing, we only apply this optimization when all input arguments are primitives and the number of input arguments is <=3.
+The figure below shows the effect of this optimization on the processing time of a query of primitive operations.
```

Comment:
maybe mention that this is actually a super common pattern in ML use cases (and not a made expression)?